File: lib/src/thunderbird.dart
import 'dart:io';

void installThunderbirdCss() {
  final profileDir = _findThunderbirdProfileDir();
  final pwd = Platform.environment['PWD'] ?? '.';

  for (final cssFileName in ['userChrome.css', 'userContent.css']) {
    /* our firefox-css is also applicable for thunderbird */
    final src = File('$pwd/assets/firefox-css/$cssFileName');
    final dest = _findChromeCssFile(cssFileName, profileDir);

    if (dest.existsSync()) dest.deleteSync();
    dest.parent.createSync(recursive: true);
    Link(dest.path).createSync(src.path);
    print('Linked ${dest.path} to ${src.path}');
  }
}

void uninstallThunderbirdCss() {
  final profileDir = _findThunderbirdProfileDir();
  for (final cssFileName in ['userChrome.css', 'userContent.css']) {
    final dest = _findChromeCssFile(cssFileName, profileDir);
    if (!dest.existsSync()) continue;
    dest.deleteSync();
    print('Deleted ${dest.path}');
  }
}

/// Finds the userChrome.css or userContent.css file for Thunderbird.
FileSystemEntity _findChromeCssFile(String cssFileName, Directory profileDir) {
  assert(cssFileName.endsWith('.css'));
  final cssFilePath = '${profileDir.path}/chrome/$cssFileName';
  final type = FileSystemEntity.typeSync(cssFilePath, followLinks: false);
  return switch (type) {
    .file => File(cssFilePath),
    .link || .notFound => Link(cssFilePath),
    _ => throw StateError('Warning: Unknown type $type for $cssFilePath'),
  };
}

Directory _findThunderbirdProfileDir() {
  final home = Platform.environment['HOME'];

  final systemProfilesDir = Directory('$home/.thunderbird');
  final flatpakProfilesDir = Directory(
    '$home/.var/app/org.mozilla.Thunderbird/.thunderbird',
  );
  final profilesDir = systemProfilesDir.existsSync()
      ? systemProfilesDir
      : flatpakProfilesDir;

  final installsIni = File('${profilesDir.path}/installs.ini');
  final installsIniContent = installsIni.readAsLinesSync();
  // Find the line with `Default=8972389472934.default-release`
  final defaultProfileId = installsIniContent
      .firstWhere((line) => line.startsWith('Default='))
      .substring('Default='.length);

  final defaultProfileDir = Directory('${profilesDir.path}/$defaultProfileId');
  if (!defaultProfileDir.existsSync()) {
    throw StateError('Could not find Thunderbird profile: $defaultProfileId');
  }

  return defaultProfileDir;
}

File: src/my_css/thunderbird.rs
use std::env::var;
use std::fs;
use std::os::unix::fs::symlink;
use std::path::{Path, PathBuf};

use anyhow::{Result, anyhow, bail};

use crate::my_css::MyCss;

impl MyCss {
  pub fn theme_thunderbird() -> Result<()> {
    assert!(Self::enabled()?);
    println!("Installing Thunderbird theme...");

    let profile_dir = Thunderbird::default_profile_dir()?;

    /* our firefox-css is also applicable for thunderbird */
    let local_dir = Path::new(&var("PWD")?).join("assets/firefox-css/");
    for subpath in ["chrome/userChrome.css", "chrome/userContent.css"] {
      let src = local_dir.join(subpath);
      let dst = profile_dir.join(subpath);

      if dst.exists() || dst.is_symlink() {
        fs::remove_file(&dst)?;
      } else {
        fs::create_dir_all(dst.parent().unwrap())?;
      }
      symlink(&src, &dst)?;
      println!("Linked {src:?} to {dst:?}");
    }

    Ok(())
  }
}

struct Thunderbird;
impl Thunderbird {
  fn default_profile_dir() -> Result<PathBuf> {
    let home = var("HOME")?;

    let legacy_profiles_dir = Path::new(&home).join(".thunderbird");
    let xdg_profiles_dir = Path::new(&home).join(".var/app/org.mozilla.Thunderbird/.thunderbird");
    let profiles_dir = if legacy_profiles_dir.exists() {
      legacy_profiles_dir
    } else {
      xdg_profiles_dir
    };

    let installs_ini = profiles_dir.join("installs.ini");
    // Find the line with `Default=8972389472934.default-release`
    let default_profile_id = fs::read_to_string(installs_ini)?
      .lines()
      .find_map(|line| line.strip_prefix("Default="))
      .ok_or_else(|| {
        anyhow!("Could not find Thunderbird profile; open Thunderbird first and try again.")
      })?
      .to_owned();

    let default_profile_dir = profiles_dir.join(format!("{default_profile_id}/"));
    if !default_profile_dir.exists() {
      bail!("Thunderbird profile {default_profile_dir:?} doesn't exist!");
    }

    Ok(default_profile_dir)
  }
}
